source/array_list.c

```diff
@@ -0,0 +1,247 @@
+        size_t last_item_offset = list->item_size * (list->length - 1);
+
+        memset((void *)((uint8_t *)list->data + last_item_offset), 0, list->item_size);
+        list->length--;
+        return AWS_OP_SUCCESS;
+    }
+
+    return aws_raise_error(AWS_ERROR_LIST_EMPTY);
+}
+
+void aws_array_list_clear(struct aws_array_list *list) {
+    if (list->length > 0) {
+        memset(list->data, 0, list->current_size);
+        list->length = 0;
+    }
+}
+
+int aws_array_list_shrink_to_fit(struct aws_array_list *list) {
+    if (list->alloc) {
+        size_t ideal_size = list->length * list->item_size;
+        if (ideal_size < list->current_size) {
+            void *raw_data = aws_mem_acquire(list->alloc, ideal_size);
+            if (!raw_data) {
+                return aws_raise_error(AWS_ERROR_OOM);
+            }
+
+            memcpy(raw_data, list->data, ideal_size);
+            aws_mem_release(list->alloc, list->data);
+            list->data = raw_data;
+            list->current_size = ideal_size;
+        }
+        return AWS_OP_SUCCESS;
+    }
+
+    return aws_raise_error(AWS_ERROR_LIST_STATIC_MODE_CANT_SHRINK);
+}
+
+int aws_array_list_copy(const struct aws_array_list *from, struct aws_array_list *to) {
+    assert(from->item_size == to->item_size);
+
+    size_t copy_size = from->length * from->item_size;
+
+    if (to->length >= from->length) {
+        memcpy(to->data, from->data, copy_size);
+        to->length = from->length;
+        return AWS_OP_SUCCESS;
+    }
+    /* if to is in dynamic mode, we can just reallocate it and copy */
+    else if (to->alloc != NULL) {
+        void *tmp = aws_mem_acquire(to->alloc, copy_size);
+
+        if (!tmp) {
+            return aws_raise_error(AWS_ERROR_OOM);
+        }
+
+        memcpy(tmp, from->data, copy_size);
+        if (to->data) {
+            aws_mem_release(to->alloc, to->data);
+        }
+
+        to->data = tmp;
+        to->length = from->length;
+        to->current_size = copy_size;
+        return AWS_OP_SUCCESS;
+    }
+
+    return aws_raise_error(AWS_ERROR_LIST_DEST_COPY_TOO_SMALL);
+}
+
+size_t aws_array_list_size(const struct aws_array_list *list) {
+    return list->current_size;
+}
+
+size_t aws_array_list_length(const struct aws_array_list *list) {
+    return list->length;
+}
+
+int aws_array_list_get_at(const struct aws_array_list *list, void *val, size_t index) {
+    if (list->length > index) {
+        memcpy(val, (void *)((uint8_t *)list->data + (list->item_size * index)), list->item_size);
+        return AWS_OP_SUCCESS;
+    }
+    return aws_raise_error(AWS_ERROR_INVALID_INDEX);
+}
+
+int aws_array_list_set_at(struct aws_array_list *list, const void *val, size_t index) {
+    if (list->current_size > index * list->item_size) {
+        memcpy((void *)((uint8_t *)list->data + (list->item_size * index)), val, list->item_size);
+
+        /* this isn't perfect but its the best I can come up with for detecting length changes*/
```

Comment:
I'm a bit confused about this. Does `set_at` implicitly extend the list? If so we need to also be prepared to resize the backing array. And this behavior needs to be documented in the header.

Reply:
My thought is that set_at should function within the existing bounds of the array (like in place replacement), and not do the resize. The semantics of having the list grow makes more sense in the push_back style interface.

Reply:
This is going to be difficult to use reliably, because whether set_at works `at all` now depends on a property of the list that 1) you have no direct control over and 2) is not likely to be examined in most use cases. It can result in code that works fine in initial testing, but then in production suddenly breaks when the list sizes align differently. It's better to have a consistent rule that method success is never tied to capacity when you have a dynamic list.

Reply:
I'll play around with it and see what happens.